### ISSUE TITLE
greylisting of whitelisted BOINC projects

### ISSUE THREAD

**Issue by Erkan-Yilmaz** [Erkan-Yilmaz/Gridcoin-tasks#6]
**fair rewarding** is important, otherwise some users get magnitude without doing anything/much.

fediverse and gridcoinstats already detect when projects have no WU, in progress WUs
but Gridcoin itself doesn't :-(

mentioned in [hangout 1](https://soundcloud.com/gridcoin-community-hangouts/gridcoin-hangout-001): 0:23:45

**Comment by grctest** [comment]
Jamezz spoke about creating a draft version of a grey-list whitepaper in the 22nd hangout.

**Comment by Erkan-Yilmaz** [comment]
cc @grcjamezz

**Comment by grcjamezz** [comment]
Thank you Erkan , since we apparently moved to bi-weekly mumbles this will give me plenty of time to write up a bit cleaner than a verbal rough draft of the concept and original idea I had along with incorporating the ideas contributed by other community members ideas all merged together along with a forum to discuss and draw up more than an idea or concept but hopefully the active mechanisms that will need to be in place both soft and hardware wise.

**Comment by Foggyx420** [comment]
i think a greylist system is a great idea and addition to the whitelist.

 Projects can run into problems or issues and a greylist could give them a certain time period to correct these issues before even a poll would be considered for removal and would solve the whole jump to a removal poll.

Perhaps maybe a greylist could temporarily block the coins for that project after a certain time point after no activity and spread them across other projects until it has moved back into the whitelist. 

Yes peoples RAC/MAG will suffer during the greylist period but its not like any work is being done for the project to begin with. gridcoin is rewarded for work and it seems whitelisted projects that run idle are just easy payouts.

just thought i would add my thoughts. cheers.

**Comment by Erkan-Yilmaz** [comment]
mentioned in [hangout 26](https://github.com/Erkan-Yilmaz/Gridcoin-hangout-minutes/blob/master/hangout_2017_03_18.MD): making a bounty for this task

**Comment by Erkan-Yilmaz** [comment]
the fair rewarding is also connected to:
* when a superblock can not gather info from a project (there are various reasons for that), the users from there do NOT get their reward, but all other Gridcoin users, see https://github.com/gridcoin/Gridcoin-Research/issues/154

**Comment by Foggyx420** [comment]
like the new title its an easier pill to swallow in a way. fair would make users unfamiliar with various subjects more accepting of the idea.

**Comment by grcjamezz** [comment]
When I mentioned the original idea it would do this #1 watch white listed projects for availiable work track and at 60 days #2 put the project in a jailed state thus the rac/mag was paused and the project was given 30 days to figure out their complications or #3 be auto removed from the whitelist.  I will use for example drug discovery , since we know they are going away I would consider right now its on the grey list... We have RAC/MAG and credit but no new WU till off the WL. If we automate the system and even put new projects on the same system so to be a trial into the gridcoin blockchain reward system if they tank in 90 days its automatic. Also 30 days should be long enough for users to move from a project with no WU to a new project and not effect their hard work to get to where they are with gridcoin. 

So I thought the gridoin team req could be used to seperate the users stats when on the 60-90 days jailed stats system ( this also would mean if someone had a system off with completed WU and sent them they would not get credit unless durring the 60-90 day period the project produced work units ).

I assume , no I know people will say a stat pause system for that 30 day period would be unfair. BUT i want to mention that lots of users dedicate themselves to 1 project trusting that if they are on boincstats/gridcoinwhitelist even a server they really is no different that apache for httpd , openssl for ssh , its a generic server with a generic front end with BOINC and BERKELEY's name all over it but Gridcoin is the best checks and balance for the BOINC system and its a bonus , you can earn a percentage of what you put in and its not 1:1 its more like 1:100 and its dynamic and we are a community make it's value. Situations where say a project is voted into the WL tanks within 6 months and you have not been rewarded for all work done due to internal issues and its voted out you never recieve that reward. Currently there is no work for a week and its on a list and 3 days latter the vote ends. This would make it asyncronus vs dynamic times/rules creating a non byist WL because someone doesnt agree with a project not having openssl or they just dont crunch it or even just see NO is the majority vote or YES and follow the majority. This creates fairness for all aspects of the internal " honored " system vs crunchers whom have been in #gridcoin for years vs the new ones in the past year , and the new ones that are just learning solo vs pool and even what the projects are about being " guided " by the trustedd #gridcoin community leaders like @Erkan-Yilmaz @grctest and @peppernrino along with the trusted help's options. We can also chart and keep track of these things most likely with netdata style loggin and tracking for work and projects over time for both historical and a good chart for new users joining leaning what works best on their hardware.


Sorry but I've only had my 2nd cup of coffee and its a saturday so laying in bed as I type this , I had not submited anything due to #gridcoin constant political and micro political bullshit. I doubt any this will be in the final project but here is some of the original idea. Thanks , take care. - Jamezz

**Comment by Mercosity** [comment]
Time to get this done.. I'm somewhat of a bystander as I don't have the relevant skillset but there is an urgent need as highlighted by the recent DD@home debacle.

**Comment by Foggyx420** [comment]
yes this would be nice to have. idk if itd just be simular to whitelist but have it rewardless?  remove project from whitelist to greylist list through simular means (i havent looked at much of that process yet). a move project feature? saving time in removal and addition between lists? just throwing ideas out there. we should brainstorm on this for sure. idk if adding an extra field on projects to say wheater its whitelisted or greylisted would be easy or not.

**Comment by tomasbrod** [comment]
Consider using total credit information from exported stats instead of RAC. Then the NN could compute magnitude based on the difference between last 2 superblocks. Also: https://github.com/gridcoin/Gridcoin-Research/wiki/DEV-Neural-Network

**Comment by tomasbrod** [comment]
Port the Neural Network to Linux first? In case the porting involves rewriting no work would be lost.
Where are the sources to DOTNET part of gridcoin?

**Comment by grctest** [comment]
We could scrape server_stats.xml for each project for data to base the grey-list on.

Example xml file: http://moowrap.net/server_status.php?xml=1

> <daemon_status> ... </daemon_status>

The above could be used to monitor the state of each BOINC project's server component, which could differentiate between a lack of WUs and a broken server.

The problem with monitoring each of these server components is that they are unique for each project, some have the default implementation where as other such as SETI have like 20 custom components. Perhaps a basic monitoring, or monitoring for 'failure/error' would be useful?

> <database_file_states> ... <database_file_states>

The contents in the above should be the primary focus of a grey listing mechanism.

Specifically the following fields within the 'database_file_states' section:
* <results_ready_to_send> = work unit availability
* <results_in_progress> = WUs current being crunched
* <results_in_progress> = No work units available, but users are awaiting validation for credit to earn Gridcoin with.
* <users_with_recent_credit> = If decreases significantly this is an additional signal of a lack of work units.

**Comment by Quezacoatl1** [comment]
I pledge 5k GRC for this cause, to give the "bounty" at least a number

**Comment by Erkan-Yilmaz** [comment]
example to comment 292554904 above:

In the current #933104 superblock are only 2 of the 25 whitelisted projects (so, 92% missing).

and the situation seems like this:
* 1000 users' rightfully earned GRCs go to 370 users (Team Users Total:1350, Team Users w/ Magnitude:373)

or in current monetary terms:
* 370 users gain now $3865.9 more in GRC

**Comment by Erkan-Yilmaz** [comment]
advise by @tomasbrod 

https://twitter.com/GridcoinIRC/status/877103074581643265

**Comment by Erkan-Yilmaz** [comment]
status on projects for last 7 days:

* Sep 11-14: CSG: [no WUs](https://steemit.com/gridcoin/@erkan/for-2-days-now-no-new-work-given-out-by-boinc-project-csg-citizen-science-grid#@erkan/re-dutch-re-erkan-for-2-days-now-no-new-work-given-out-by-boinc-project-csg-citizen-science-grid-20170916t063415519z) for 3.5 days
* Sep 11: Asteroids: [maintainance](https://steemit.com/gridcoin/@erkan/planned-maintainance-tomorrow-boinc-project-asteroids-home-for-5-7-days-sep-11-18)
  * Sep 17: [no WUs](https://steemit.com/gridcoin/@erkan/planned-maintainance-tomorrow-boinc-project-asteroids-home-for-5-7-days-sep-11-18#@erkan/re-erkan-planned-maintainance-tomorrow-boinc-project-asteroids-home-for-5-7-days-sep-11-18-20170918t051150214z) - investigation in progress still
* Sep 13: Cosmology, [no WUs for about 1 day](https://steemit.com/gridcoin/@erkan/available-work-less-zero-at-boinc-project-cosmology-home#@erkan/re-dutch-re-erkan-available-work-less-zero-at-boinc-project-cosmology-home-20170916t063946495z)

**Comment by Erkan-Yilmaz** [comment]
a very, quick explanation of this, from IRC:

user1: i am not understand this all with the "greylistings"
* me: the simplest case in the greylisting idea (read more there) is: when user X bunkers lots of WUs, and other users can NOT get WUs anymore. Make a guess which user will get GRCs ?
* me: this would be on purpose making zero WUs
* me: then there are unplanned events how zero WUs can come to be

user1: in boinc max bunker volume are 5 + 5 days
* me: you can prepare a few accounts, or how do you think the WUs got so fast empty in Pentathlon or Formula BOINC challenges ?
* me: it's simple to do and is done by lots of BOINC teams

**Comment by G-UK** [comment]
This is now completed, I put together a proposal that was accepted and is now in operation.

https://github.com/gridcoin-community/Gridcoin-Tasks/issues/194

@Erkan-Yilmaz

**Comment by 47an** [comment]
I add 2000 GRC on this and a fair share of used contribution used to code on this.
(these coins would be send and fill up on existing bounty)

**Comment by Quezacoatl1** [comment]
I agree that we now have a greylist. Thought when I pledged my part of the bounty, I thought about an automated system in the code.

It first crossed my mind to split the bounty for the current greylist and the new, automated greylist. But that is too complicated. So I think at least my bounty should go to the creators of the current system including the website(s?) That is displaying the greylist. :) 
As I do not know a complete list of actors who participated, I would like to get some input on how to divide the bounty.

**Comment by dopeshitnetworks-irc-dopeshit-net** [comment]
Quezaciatk1 , that seems to be debatable and honestly I personally think that payment for what currently exists does not qualify for what was intended when a bounty was created. I fully agree as I intended it to be automated vs editing a .php website manually. I consider the project not accomplished or under " not at the level of quality expected by the community " and it should not be accepted as completed. 
I know personally just conversing via the hangouts starting @ #22  when I first brought up the idea in January of 2017 and had to first get others on board with the idea and convincing customminer that I actually came up with a good idea and contribution earning the backing from a chunk of the  mumble/irc active members of the community along with some here on github along with CCT to take it from just an idea in my head to working out details being as I stated previously 40+ hours of hangout time including after the hangouts and during the week running my thoughts and ideas past peers since I was not trying to be a lone wolf and implement something no different than the current wiki whitelist page but pulling the projects from boincstats list of all projects and just adding " greylist " after a project name even if it was never whitelisted nor would ever be under consideration along with projects that have been blacklisted and removed such as drugdiscovery or bitcoin utopia. I know G_UK is going to puff out his chest and claim it was his idea as he did when he created the poll taking full credit for thinking up the idea let alone the cheapskate making it as if it was a school project made by a middle school child and using dyndns vs it being incorporated into www.gridcoin.us replacing the cname dns forward currently pointed to wikipedia  for www.gridcoin.us/current_whitelist or what not. I had actually conversed how to automate the process and fully believe it should be automated and hands off and non manipulatable by a human. 
This task is far from complete , anybody can throw up a webpage and manually add and delete text but being taken from the Gridcoin Neural Network it eliminates user error.
I am not going to cry if I do not receive anything for my time and effort , but I am sure G_UK will as the bounty is the most likely the only reason he setup his K.I.S.S. based site with as little effort as possible. Again he states it was his idea.. Interesting that the notes from the hangouts and recordings prove otherwise.

**Comment by 47an** [comment]
I´m up for make it automated before close.

thanks for post jamezz, i would like to see what you have in your mind to contribute to make this complete.

**Comment by tomasbrod** [comment]
I do not agree that @G-UK put as little effort as possible into his whitelist website.
He put more.
